src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,230 @@
 package core.basesyntax;
 
+import java.util.Collection;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> firstNode;
+    private Node<T> lastNode;
+    private int size;
+
     @Override
     public void add(T value) {
+        linkedLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            linkedLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
+    }
+
+    void linkBefore(T e, Node<T> succ) {
+        final Node<T> pred = succ.previous;
+        final Node<T> newNode = new Node<>(pred, e, succ);
+        succ.previous = newNode;
+
+        if (pred == null) {
+            firstNode = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    public void addAll(int index, Collection<? extends T> c) {
+        checkPositionIndex(index);
+
+        Object[] a = c.toArray();
+        int numNew = a.length;
+        if (numNew == 0) {
+            return;
+        }
+
+        Node<T> predecessor;
+        Node<T> successor;
+        if (index == size) {
+            successor = null;
+            predecessor = lastNode;
+        } else {
+            successor = node(index);
+            predecessor = successor.previous;
+        }
+
+        for (Object o : a) {
```

Comment:
Don't create repeating code. You can avoid code duplication by creating a method like `findNodeByIndex` and `checkIndex` to handle these repeated logic segments.